### PR TITLE
adds System.register output

### PR DESCRIPTION
**Pull request by robwormald**
This adds a single file build in the System.register format. I'm not entirely sure on the deployment step, so this isn't merge ready yet. See http://plnkr.co/edit/rTJV9QzyHMVfnKNynqKu?p=preview for usage.

TODOs that I might need assistance with:
- pushing the built angular2.js to a CDN
- adding angular2 to the jspm registry.

See http://plnkr.co/edit/rTJV9QzyHMVfnKNynqKu?p=preview for demo.

### CODE REVIEW

**Comment by robwormald** [comment]
Updated the CLA with my Github username.

File: gulpfile.js [hunk 1]
```diff
@@ -643,7 +646,18 @@ gulp.task('build.js.cjs', function(done) {
   );
 });
 
-gulp.task('build.js', ['build.js.dev', 'build.js.prod', 'build.js.cjs']);
+gulp.task('build.js.system',['build.js.dev'],function(){
```
Comment (rkirov) [inline gulpfile.js]: space after each comma.

In general try to follow basic https://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml rules.

File: package.json [hunk 2]
```diff
@@ -71,10 +71,29 @@
...
+  "jspm": {
+    "main": "modules/angular2/angular2.js",
+    "directories": {
+      "lib": "modules/angular2",
+      "packages": "jspm_packages"
+    },
+    "configFile": "jspm.conf.js",
+    "dependencies": {
+      "es6-module-loader": "^0.9.2",
+      "googleapis": "1.0.x",
```
Comment (rkirov) [inline package.json]: What is the meaning of this dependency list? It looks too broad for runtime `system` dependencies, and for build dependencies we already have `npm` dependencies.

**Comment by zzo** [comment]
Another way to do this is to just add the 'moduleName: true' option to es5build.js - maybe less invasive?  Altho you would need to customize the loader using something like:

```
    var systemNormalize = System.normalize;
    System.normalize = function(name, parentName, parentAddress) {
      return Promise.resolve(systemNormalize.call(this, name, parentName, parentAddress)).then(function(normalized) {
          if (normalized && normalized.match(/^angular2/)) {
            normalized += '.js';
          }
          return normalized;
      });
    };
```
Because there's no way (yet?) to dynamically configure the moduleName using gulp-traceur (If we could strip off the '.js' off the filename & use that as the moduleName I believe both loading methods would work (loading one concatenated file or all of them individually) from the same set of transpiled files.
There is a PR against gulp-traceur for this already: https://github.com/sindresorhus/gulp-traceur/issues/54

Just throwing that out there although I like this solution too but this has a lot more going on!  Regardless Big +1 to this feature.

File: package.json (hunk 2 above)
Comment (robwormald) [inline package.json]: Ah, got that wrong moving things around, corrected.

In theory, these could be moved into the root of the package.json, but I didn't want to get overzealous - JSPM gives you the option of prefixing so I thought I'd start there and get some input.

Usage for end users is/would be:
`jspm init` 
`jspm install angular2`

**Comment by robwormald** [comment]
The basic issue I had is that because there's a mix of .es6 and .js extensions in the source, its difficult to specify which to use without hardcoding a bunch of paths again. So currently the task is just tailing off the existing buildsteps, as that gives a directory full of .es6 files. This whole task could probably be eliminated once everything normalizes a bit, the idea was to get a single-file build for demos and plunkers. 

We've been using SystemJS/JSPM for a few months now, and I just can't highly recommend it enough.

**Comment by zzo** [comment]
Thanks Hope you don't mind a couple questions 😄

- is it transpiling the es6 sources directly in the browser via traceur?

- jspm seems to say it can follow only the dependencies needed will that be possible/ how will that look instead of bundling up everything?

- if I'm only using 'instantiate' type modules can I just use the es6 loader and the system.register extension instead of pulling in all of systemjs using this (like I'm currently doing)

-sorry if I didn't catch this before but is there any jspm source in the final bundle or is it just full traceur zonejs es6-module-loader and systemjs at the top and the concatenated es6 source in order at the bottom?

THANKS for doing this and helping me understand it!!

**Comment by robwormald** [comment]
@zzo happy to answer questions!

In this case, no - It's using the SystemJS builder to compile into a single bundle of ES5 code (In the System.register format). It certainly *can* compile on the fly (this is how we do most of our development), but the goal of this was to get a single file for ease of use in demos. 

The above System.register format, when included in a script tag, registers built modules with the System loader. This means, it'll check to see if a dependency is loaded before loading it remotely.

The angular2.js output is just angular code, without anything else bundled. IMO, this is the best way going forward.

**Comment by ajoslin** [comment]
The problem with this is it only compiles the files which are referenced from `angular2/angular2.js`.

Which means that importing things like this fail:

```js
import {PrivateComponentLocation} from 'angular2/src/core/compiler/private_component_location';
```

This may be a problem on angular's end ... everything that is public and documented (like PrivateComponentLoader) should be exposed in the angular2.js file.

The build system should, though, at the end of the day compile `**/*.es6`, not just `angular.es6`.

**Comment by zzo** [comment]
Thanks @robwormald  - so the user must load traceur_runtime.js and es6-module-loader (& systemjs?) before loading this generated angular2.js bundle? thanks!

**Comment by robwormald** [comment]
@ajoslin yeah - the ES6 loader stuff will track down any *referenced* dependencies, and bring them in. If they're not referenced anywhere, they'll get ignored (which is actually great, it's a bit like Dart's tree-shaking that way). 

I think the real question is how end users will use this stuff when building their own apps. I'd much prefer to bring in the raw At/Typescript, import what I want, and then bundle all my code into a single build. 

Dumping out a bundle for demo purposes is great, but I think it sorta defeats the purpose for end-users. This is why I'm so hot on having JSPM as part of the workflow, because that's what it's set up with in mind.

**Comment by robwormald** [comment]
Just to clarify - I'm not in any way advocating JSPM/SystemJS be a requirement. It's one of many options. 
- AMD users will want to consume ES5
- CommonJS users will (probably?) want ES5
- "Pure" ES6 users will want raw ES6 source (probably) to be bundled in with their own processes. 
- I'm entirely unsure how ES5 users will consume this repo. 

For reference, at work, we have a huge amount of ES6 'service' type code, which we build and bundle into an SFX ES5 bundle, and then bring that into Angular. It works well, but it causes some complications when mixed with ES5, minified, etc. 

For ES6 apps, my feeling is its best to keep discrete files until the user is ready to bundle them.

**Comment by robwormald** [comment]
If you use the http://jspm.io/system.js CDN link, it will bring them both in automatically. 

Otherwise yes, you'd need traceur-runtime and the es6 loader. SystemJS adds the ability to load any format (cjs/amd) as well - it's a universal loader.

**Comment by pkozlowski-opensource** [comment]
@robwormald would you mind rebasing and squasing all the commits? I would like to play with the produced output and see what gets produced, but your PR doesn't merge cleanly any more and we had some new, mandatory dependencies introduced recently (rx.js).

Will this build make sure that zone.js scripts are included at the very begining of the produced file?

**Comment by robwormald** [comment]
@pkozlowski-opensource sure, i'll try to do it on my lunch break.

re: ordering - 
- Zone
- Traceur Runtime
- SystemJS/ES6-loader
- RX
- Angular

Sound right? Do we want to bundle RxJS, or have it included as an external dep?

**Comment by pkozlowski-opensource** [comment]
@robwormald I'm not sure we should be including system loader, on a fence here. For rx - I don't think we've got a choice, it is a mandatory dependency as far as I can tell.

**Comment by robwormald** [comment]
Roger. I have a bundle up @ http://code.innit.io/angular2/angular2.js and also http://code.innit.io/angular2/angular2.dev.js - that's just ng + rx if you want to have a looksee.

File: package.json [hunk 3]
```diff
@@ -74,10 +74,22 @@
     "run-sequence": "^0.3.6",
     "source-map": "^0.3.0",
     "sprintf-js": "1.0.*",
+    "systemjs-builder": "git://github.com/systemjs/builder",
```
Comment (pkozlowski-opensource) [inline package.json]: Why installing from git? Is there anything wrong with the version [deployed to npm](https://www.npmjs.com/package/systemjs-builder)?

File: package.json [hunk 4]
```diff
@@ -74,10 +74,22 @@
...
     "vinyl": "^0.4.6",
     "yargs": "2.3.*"
+  },
+  "jspm": {
+    "main": "modules/angular2/angular2.js",
+    "directories": {
+      "lib": "modules/angular2",
+      "packages": "jspm_packages"
+    },
+    "configFile": "jspm.conf.js",
```
Comment (pkozlowski-opensource) [inline package.json]: I can't see `jspm.conf.js` anywhere in this PR... Is this `jspm` config needed / used at all?

File: package.json [hunk 5]
```diff
@@ -74,10 +74,22 @@
...
+    "dependencies": {
+      "angular/zone.js": "github:angular/zone.js@^0.4.1"
```
Comment (pkozlowski-opensource) [inline package.json]: So, zones shouldn't be a System.import dependency but rather a file whose content is pre-appended "manually"

**Comment by pkozlowski-opensource** [comment]
@robwormald I was trying to use the build from this PR but bumped into few problems that might be linked to my env on this particular machine - going to try it on a different OSX installation later today. But I went over the diff and left some comments.

Will try to ping you on gitter later on

**Comment by rkirov** [comment]
This PR achieves two goals:
* creates a gulp task that produce a System.js using `system-builder`.
* adds jspm annotations to `package.json` to make `angular2` a jspm cli consumable.

The first task is a must and of high-priority. You are on the right track there, one thing we want to add is prepending with `zone.js` for even more friendly package.

For the second we are still evaluating `jspm` vs competing browser dep management solutions and not ready to commit. It would be best if the two goals are in separate PRs, so that we can land the first one faster as we are debating the second one.

File: .gitignore [hunk 6]
```diff
@@ -4,6 +4,7 @@ packages/
 .buildlog
 node_modules
 bower_components
+jspm_packages
```
Comment (pkozlowski-opensource) [inline .gitignore]: is this one still needed?

**Comment by pkozlowski-opensource** [comment]
@robwormald thnx for all the fixes, it starts to look like sth we can get in. Would you mind squasing all the commits into one, rebasing and making sure that the TravisCI build is passing?

**Comment by pkozlowski-opensource** [comment]
Thnx!

**Comment by alexeagle** [comment]
Can we add some sort of e2e test that ensures the resulting bundles actually work?

I am working to migrate from traceur to typescript. This de-stabilizes the structure of the tasks which do the transpilation and I am not sure how to verify if I am breaking the distributed bundle.

**Comment by pkozlowski-opensource** [comment]
@alexeagle definitively, this is part of the whole packaging story plan. But for now @rkirov and I are focusing on getting the content of those packages right. But yeh, e2e tests are must.

**Comment by robwormald** [comment]
@alexeagle happy to do some E2E if its possible. 

In this first rev, I tried to be as non-invasive as possible, so the bundler is running on post-compiled ES6 file (from dist/js/prod or dist/js/dev) - it simply accepts the 'entry point' and uses the es6 imports to pull in everything else. Its set up to run after the build tasks, so as long as the dist location remains the same, it should be fairly stable... looking at it, I'll tweak the bundler code to use the CONFIG paths in the gulpfile to try to keep stuff unified. 

The right way to do this would be bundling directly from source, but that's a bit impossible at the moment with the mixed .es6 and .js source files (and presumably soon to be .ts...) 

They are tracking systemjs support in TS over here https://github.com/Microsoft/TypeScript/issues/2233, using the jspm-typescript plugin.

There's a further option I'm not leveraging in the PR, called an "SFX" bundle, which outputs a dependency-free es5 build bundled with a microloader - we currently use this workflow to build our SDK at work and then bring that into angular1 - it doesn't need the SystemJS loader included, so this might be a good option for es5 consumers.

**Comment by robwormald** [comment]
@pkozlowski-opensource squashed

**Comment by rkirov** [comment]
merged.

**Pull request closed without merging**